lib/pushmi_pullyu/cli.rb

```diff
@@ -32,39 +32,37 @@ def run
 
   def start_server
     setup_signal_traps
+
     setup_log
     print_banner
 
-    begin
-      run_tick_loop
-    rescue Interrupt
-      logger.info 'Shutting down'
-      @running = false
-      logger.info 'Bye!'
-      exit(0)
-    end
+    run_tick_loop
   end
 
   private
 
   def setup_signal_traps
-    Signal.trap('INT') { raise Interrupt }
-    Signal.trap('TERM') { raise Interrupt }
+    Signal.trap('INT') { Thread.new { shutdown } }
+    Signal.trap('TERM') { Thread.new { shutdown } }
     Signal.trap('HUP') do
-      if config.logfile
-        logger.debug 'Received SIGHUP, reopening log file'
-        # TODO: reopen logs
-        # PushmiPullyu::Logging.reopen_logs(config.logfile)
+      Thread.new do
+        PushmiPullyu::Logging.reopen
       end
     end
   end
 
+  def shutdown
```

Comment:
This allows for a graceful shutdown, by stopping the run loop and letting the program end itself. If you want to force shutdown, issue another INT/TERM to the program (CTRL-C or `pushmi_pullyu stop`, etc)